### PR TITLE
Detect missing `;` on methods with return type `()`

### PR DESCRIPTION
**Pull request by estebank**
On a given file `foo.rs`:

``` rust
fn foo() {
    return 1;
}

fn main() {
    3
}
```

Provide the following output:

``` bash
error[E0308]: mismatched types
 --> foo.rs:2:12
  |
2 |     return 1;
  |            ^ expected (), found integral variable
  |
  = note: expected type `()`
  = note:    found type `{integer}`

error[E0308]: mismatched types
 --> foo.rs:6:5
  |
6 |     3
  |     ^ expected (), found integral variable
  |
  = note: expected type `()`
  = note:    found type `{integer}`
help: maybe you meant to have a statement ending in `;` here
 --> foo.rs:6:6
  |
6 |     3
  |      ^

error: aborting due to 2 previous errors
```

Fixes: #25133

### CODE REVIEW

**Comment by jonathandturner** [comment]
Thanks for the PR!

You know, in looking at this I wonder if we can combine the two.  Something like:

```bash
error[E0308]: mismatched types
 --> foo.rs:6:5
  |
6 |     3
  |     ^ expected (), possibly missing `;` at end
  |
  = note: expected type `()`
  = note:    found type `{integer}`
```

**Comment by jonathandturner** [comment]
Or maybe

```bash
error[E0308]: mismatched types
 --> foo.rs:6:5
  |
6 |     3
  |     ^ expected type `()`, missing `;` at end?
  |
  = note: expected type `()`
  = note:    found type `{integer}`
```

**Comment by nikomatsakis** [comment]
One last possibility might be

```
error[E0308]: mismatched types
 --> foo.rs:6:5
  |
6 |     3
  |     ^- possibly missing `;` here?
  |     expected type `()`, missing `;` at end?
  |
  = note: expected type `()`
  = note:    found type `{integer}`
```

that is, add another label -- of course this runs afoul of the case where our rendering looks wacky.

**Comment by jonathandturner** [comment]
Of the three I vote for https://github.com/rust-lang/rust/pull/36409#issuecomment-246400438, but take your pick :)

**Comment by estebank** [comment]
@jonathandturner I'll use that one,  seems to be the least verbose of the options. I'll update this later today.

**Comment by estebank** [comment]
@jonathandturner I went with @nikomatsakis' style as it lend itself to a cleaner diff and didn't look half bad :)

```bash
error[E0308]: mismatched types
 --> unit-error.rs:6:5
  |
6 |     3
  |     ^
  |     |
  |     possibly missing `;` here?
  |     expected (), found integral variable
  |
  = note: expected type `()`
  = note:    found type `{integer}`
```

**Comment by jonathandturner** [comment]
nit: I think it reads slightly better in the opposite order

```
error[E0308]: mismatched types
 --> unit-error.rs:6:5
  |
6 |     3
  |     ^
  |     |
  |     expected (), found integral variable
  |     possibly `;` missing here?
  |
  = note: expected type `()`
  = note:    found type `{integer}`
```

**Comment by jonathandturner** [comment]
Actually, scratch that, either way is fine

**Comment by estebank** [comment]
@nikomatsakis 
>that is, add another label -- of course this runs afoul of the case where our rendering looks wacky.

I think it doesn't look half bad in this case:

```
error[E0308]: mismatched types
  --> /Users/ekuber/personal/rust/src/test/compile-fail/issue-20862.rs:12:5
   |
12 |     |y| x + y
   |     ^^^^^^^^^
   |     |
   |     possibly `;` missing here?
   |     expected (), found closure
```

---

@jonathandturner I'm puzzled by the test failures. For the test `compile-fail/issue-20862.rs`, `runtest.rs` on this branch sees the following errors being emitted:

```
[
Error { line_num: 12, kind: Some(Error), msg: "12:5: 12:14: mismatched types [E0308]" },
Error { line_num: 12, kind: Some(Error), msg: "12:5: 12:14: mismatched types [E0308]" },
Error { line_num: 12, kind: Some(Note), msg: "possibly `;` missing here?" },
Error { line_num: 12, kind: Some(Note), msg: "expected (), found closure" },
Error { line_num: 12, kind: Some(Note), msg: "12:5: 12:14: expected type `()`" },
Error { line_num: 12, kind: Some(Note), msg: "12:5: 12:14: expected type `()`" },
Error { line_num: 12, kind: Some(Note), msg: "12:5: 12:14:    found type `[closure@/Users/ekuber/personal/rust/src/test/compile-fail/issue-20862.rs:12:5: 12:14 x:_]`" },
Error { line_num: 12, kind: Some(Note), msg: "12:5: 12:14:    found type `[closure@/Users/ekuber/personal/rust/src/test/compile-fail/issue-20862.rs:12:5: 12:14 x:_]`" },
Error { line_num: 17, kind: Some(Error), msg: "17:13: 17:22: expected function, found `()`" }]
```

for the following compiler visual output:

```
error[E0308]: mismatched types
  --> /Users/ekuber/personal/rust/src/test/compile-fail/issue-20862.rs:12:5
   |
12 |     |y| x + y
   |     ^^^^^^^^^
   |     |
   |     possibly `;` missing here?
   |     expected (), found closure
   |
   = note: expected type `()`
   = note:    found type `[closure@/Users/ekuber/personal/rust/src/test/compile-fail/issue-20862.rs:12:5: 12:14 x:_]`

error: expected function, found `()`
  --> /Users/ekuber/personal/rust/src/test/compile-fail/issue-20862.rs:17:13
   |
17 |     let x = foo(5)(2);
   |             ^^^^^^^^^

error: aborting due to 2 previous errors
```

While in master for the same file, the errors emitted are:
```
[
Error { line_num: 12, kind: Some(Error), msg: "12:5: 12:14: mismatched types [E0308]" }, 
Error { line_num: 12, kind: Some(Note), msg: "expected (), found closure" }, 
Error { line_num: 12, kind: Some(Note), msg: "12:5: 12:14: expected type `()`" }, 
Error { line_num: 12, kind: Some(Note), msg: "12:5: 12:14:    found type `[closure@/Users/ekuber/personal/rust/src/test/compile-fail/issue-20862.rs:12:5: 12:14 x:_]`" }, 
Error { line_num: 17, kind: Some(Error), msg: "17:13: 17:22: expected function, found `()`" }
]
```

That made me think that I might have introduced a double `emit()` call to the same `DiagnosticBuilder` somewhere in my PR, but couldn't find it.

By commenting out `src/librustc/infer/error_reporting.rs:554-558` (the new `span_label()` call), the emitted errors are the same as in master, which leads me to believe that there might be some quirk in the handling of multiple labels on the same span.

If this is indeed the case, does it sound ok if I work around it for now by duplicating the expected error messages in the affected unittests and file a ticket to fix this (either by changing how the emitter works or by introducing deduplication in `runtest.rs`)?

---

Also, `compile-fail/issue-20862.rs` shows a situation where this help message would be wrong, as the problem was with the original `fn` definition lacking the return type. Is it ok if I leave the effort to check the usages for that possibility for a different PR?

File: src/librustc/infer/error_reporting.rs [hunk 1]
```diff
@@ -523,7 +523,8 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                          origin: TypeOrigin,
                          secondary_span: Option<(Span, String)>,
                          values: Option<ValuePairs<'tcx>>,
-                         terr: &TypeError<'tcx>)
+                         terr: &TypeError<'tcx>,
+                         implicit_return: bool)
```
Comment (eddyb) [inline src/librustc/infer/error_reporting.rs]: Could you rename this, and all other instances, to `is_block_tail`? It doesn't seem to be related to `return` at all.

Comment (estebank) [inline src/librustc/infer/error_reporting.rs]: @eddyb, renamed the field as requested.

File: src/librustc/infer/error_reporting.rs [hunk 2]
```diff
@@ -540,10 +541,11 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         let span = origin.span();
 
         if let Some((expected, found)) = expected_found {
-            let is_simple_error = if let &TypeError::Sorts(ref values) = terr {
-                values.expected.is_primitive() && values.found.is_primitive()
+            let (is_simple_error, expected_unit) = if let &TypeError::Sorts(ref values) = terr {
+                (values.expected.is_primitive() && values.found.is_primitive(),
+                    &format!("{}", values.expected) == "()")
```
Comment (eddyb) [inline src/librustc/infer/error_reporting.rs]: Doesn't `.is_unit()` (or I suppose `.is_nil()`) work here?

Comment (estebank) [inline src/librustc/infer/error_reporting.rs]: `is_nil` seems like it would the appropriate method. Running the tests now.

File: src/librustc_typeck/check/demand.rs [hunk 3]
```diff
@@ -58,7 +59,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         if let Err(e) = self.try_coerce(expr, checked_ty, expected) {
             let origin = TypeOrigin::Misc(expr.span);
             let expr_ty = self.resolve_type_vars_with_obligations(checked_ty);
-            self.report_mismatched_types(origin, expected, expr_ty, e);
+            let parent_node = self.tcx.map.get(self.tcx.map.get_parent_node(expr.id));
+            let implicit_return = match parent_node {
+                Node::NodeBlock(_) => true,
+                _ => false,
+            };
+            self.report_mismatched_types_for_return(origin, expected, expr_ty, e, implicit_return);
```
Comment (eddyb) [inline src/librustc_typeck/check/demand.rs]: Hmm, thinking about it more, could the logic be moved here? Or even better, in `check_block_with_expectation`?
See #37412 for how I moved a note there. Maybe I should get that PR merged first, would make this easier.

**Comment by brson** [comment]
I question the premise of this PR.

There's been a lot of debate here about how to tell the user to put a semicolon at the end of this expression. But in this example, _ putting a semicolon after the literal "3" is not correct_. It results in nonsense code.

Maybe we should think deeper about what might be going wrong when a block that is supposed to be terminated with `()` is not. Is it _ever_ correct to just add a semicolon? I'm not sure, but I'd guess that is less likely to be the right solution than others because it's rare in Rust to not use a return value, so any value produced by the trailing expression is probably intended to be used, not thrown away.

**Comment by nikomatsakis** [comment]
@brson it happens to be pretty regularly that the right answer **is** to add a semicolon. Admittedly, not with an expression like `3`. Much more common is something like `map.insert(key, value)`. I think it will depend a lot on the kind of expression.

Two caveats though:

1. A lot of times, this happens in closure expressions, like `foo.bar(|elem| map.insert(elem, 1))`,  in which case I have to change it to `foo.bar(|elem| { map.insert(elem, 1); })`, so just adding a semicolon is not enough.
2. The other **very** common case for me personally is that I forgot to write the return type, most commonly in `fn new(...) { SomeStruct { ... } }`.

This last example again suggests that we might examine the kind of expression to see whether to suggest *changing the return type* or *adding a semicolon*.

**Comment by estebank** [comment]
> @brson it happens to be pretty regularly that the right answer is to add a semicolon. Admittedly, not with an expression like 3. Much more common is something like map.insert(key, value). I think it will depend a lot on the kind of expression.

I agree with this (unsurprisingly, as I wrote the PR in the first place).

> This last example again suggests that we might examine the kind of expression to see whether to suggest changing the return type or adding a semicolon.

I was just thinking about that very same case, and I think that a reasonable approach to this would be to point out both options, given that we cannot really find a single recommendation that will be right 100% of the time:

```
error[E0308]: mismatched types
 --> foo.rs:6:5
  |
5 | fn foo() {
  | -------- possibly return type `{integer}` missing here?
6 |     3
  |     ^
  |     |
  |     expected (), found integral variable
  |     possibly `;` missing here?
  |
  = note: expected type `()`
  = note:    found type `{integer}`
```

**Comment by jonathandturner** [comment]
@estebank - yeah I'm okay with that.  An alternative would be to use a suggestion:

```
error[E0308]: mismatched types
 --> foo.rs:6:5
  |
5 | fn foo() {
  | -------- possibly return type `{integer}` missing here?
6 |     3
  |     ^
  |     |
  |     expected (), found integral variable
  |
  = note: expected type `()`
  = note:    found type `{integer}`
suggestion:
  |
6 |     3
  |      ^ possibly missing ';' here
  
```

**Comment by estebank** [comment]
I've been looking around and by the time the type error is raised I have no longer access to any `NodeId`, which I need to find the `Span` for `fn foo()`. Is it ok if we merge this without that feature while I try to refactor the code to allow for that suggestion?

**Comment by eddyb** [comment]
@estebank I believe that if you follow https://github.com/rust-lang/rust/pull/36409#discussion_r86723295 you'll regain that ability quite readily.
I don't want to have this merged in this state if it can be simpler (as it will interact with other changes).

**Comment by GuillaumeGomez** [comment]
Just to say:

> possibly `;` missing here?

👎 

> possibly missing ';' here?

👍

**Comment by estebank** [comment]
@eddyb, Oh, my. I completely missed that comment when I got back to this PR. Just pushed a change implementing that recommendation and adding the method return type suggestion. I still need to work on this to avoid suggesting _adding_ a return type when one is already provided, maybe suggesting _changing_ it to the supplied type, and not to suggest anything when it is implementing a trait.

**Comment by estebank** [comment]
@GuillaumeGomez done.

File: src/librustc/infer/mod.rs [hunk 4]
```diff
@@ -1447,14 +1447,23 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
...
         let trace = TypeTrace {
             origin: origin,
             values: Types(ExpectedFound {
                 expected: expected,
                 found: actual
             })
         };
-        self.report_and_explain_type_error(trace, &err).emit();
+        self.report_and_explain_type_error(trace, &err)
     }
```
Comment (eddyb) [inline src/librustc/infer/mod.rs]: If you added this function you should use it wherever `report_and_explain_type_error` was used for the same effect.

File: src/librustc_typeck/check/demand.rs [hunk 5]
```diff
@@ -58,7 +60,34 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         if let Err(e) = self.try_coerce(expr, checked_ty, expected) {
             let origin = TypeOrigin::Misc(expr.span);
             let expr_ty = self.resolve_type_vars_with_obligations(checked_ty);
-            self.report_mismatched_types(origin, expected, expr_ty, e);
+            let mut diag = self.mismatched_types_diag(origin, expected, expr_ty, e.clone());
+
+            if let Node::NodeBlock(block) = self.tcx.map.get(self.tcx.map.get_parent_node(expr.id)) {
+                if let TypeError::Sorts(ref values) = e {
```
Comment (eddyb) [inline src/librustc_typeck/check/demand.rs]: You can move this to `check_block_with_expected` and it will be simpler.

File: src/librustc_typeck/check/demand.rs [hunk 6]
```diff
@@ -58,7 +60,34 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                        // an implicit return to a method with return type `()`
+                        diag.span_label(expr.span,
+                                        &"possibly missing `;` here?");
+                        // get the current node's method definition
+                        if let Node::NodeExpr(item) = self.tcx.map
+                            .get(self.tcx.map.get_parent_node(block.id))
+                        {
+                            if let Node::NodeItem(func) = self.tcx.map
+                                .get(self.tcx.map.get_parent_node(item.id))
+                            {
```
Comment (eddyb) [inline src/librustc_typeck/check/demand.rs]: I don't like this, it's too arbitrary. If you want to do this properly you'd have to add something to `Expectation` to indicate that the type comes from a function return type.

File: src/test/compile-fail/block-must-not-have-result-do.rs [hunk 7]
```diff
@@ -10,6 +10,14 @@
...
 fn main() {
     loop {
-        true //~  ERROR mismatched types
+        true
+        //~^ ERROR mismatched types
+        //~| ERROR mismatched types
+        //~| NOTE: possibly missing `;` here?
+        //~| NOTE: expected (), found bool
+        //~| NOTE: expected type `()`
+        //~| NOTE: expected type `()`
```
Comment (estebank) [inline src/test/compile-fail/block-must-not-have-result-do.rs]: I didn't dig too deeply into this [when I noticed it](https://github.com/rust-lang/rust/pull/36409#issuecomment-250346774), but we should probably create a ticket for it. Could the same DiagnosticBuilder be emitter multiple times? Maybe that's what's happening...

**Comment by estebank** [comment]
Output for `file.rs`:

```rust
fn foo(a: u32, b: u32) {
        a + b
}

fn bar(a: u32, b: u32) -> () {
        a + b
}

fn qux(a: u32, b: u32) {
    3
}

fn main() {
    3
}
```
new compiler output

**Comment by eddyb** [comment]
@estebank @jonathandturner Would it be possible to put the label about `;` just after the expression?

**Comment by eddyb** [comment]
Btw, two of my comments (https://github.com/rust-lang/rust/pull/36409#discussion_r87774028 and https://github.com/rust-lang/rust/pull/36409#discussion_r87774419) appear to still be relevant.

**Comment by estebank** [comment]
> Would it be possible to put the label about ; just after the expression?

I believe that I can create a span right after the last char for the expression.

> Btw, two of my comments (#36409 (comment) and #36409 (comment)) appear to still be relevant.

You're right, I didn't have the time to incorporate them over the weekend, just to fix the merge conflict and remove unused code. I will be doing so within the next few days.

File: src/librustc_typeck/check/demand.rs (hunk 6 above)
Comment (estebank) [inline src/librustc_typeck/check/demand.rs]: Take a look at the [latest code](https://github.com/rust-lang/rust/pull/36409/files#diff-1d1b0d29a2e8da97c6bfb6e364d920c7R4101) , where I am still against the block's parent to see wether it is a function or a trait's method for the recommendation. Let me know if it still looks too out of place.

I also changed the label to be consistent with the existing phrasing "consider removing this semicolon"/"consider adding a semicolon here".

File: src/librustc_typeck/check/mod.rs [hunk 8]
```diff
@@ -4085,7 +4086,57 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                            hi: e.span.hi + BytePos(1),
+                            expn_id: e.span.expn_id
+                        };
+                        err.span_label(span, &"consider adding a semicolon here");
+                    }
+
+                    // Is the block part of a fn?
+                    let parent = self.tcx.map.get(self.tcx.map.get_parent(blk.id));
+                    let fn_decl = if let Node::NodeItem(&hir::Item {
+                        name, node: hir::ItemFn(ref decl, ..), ..
```
Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: So this is the only part I'm wary about anymore: doing it based *solely* on syntactical nesting is bound to be either too specific, or wrong. The only way I can see this working as intended is if `ExpectHasType` carried the extra information that it corresponds to a `()` implicit return type. cc @rust-lang/compiler

**Comment by steveklabnik** [comment]
Closing due to inactivity; @estebank if you end up rebasing this, please let us know and we can re-open!
